app/internal/utils.py

```diff
@@ -57,3 +59,25 @@ def get_user(session: Session, user_id: int) -> Optional[User]:
         None
     """
     return session.query(User).filter_by(id=user_id).first()
+
+
+def get_time_from_string(string: str) -> Optional[Union[date, time]]:
+    """Converts time string to a date or time object.
+
+    Args:
+        string (str): Time string.
+
+    Returns:
+        datetime.time | datetime.date | None: Date or Time object if valid,
+                                              None otherwise.
+    """
+    try:
+        return datetime.strptime(string, '%Y-%m-%d').date()
+    except ValueError:
+        try:
+            return datetime.strptime(string, '%H:%M').time()
+        except ValueError:
+            try:
+                return datetime.strptime(string, '%H:%M:%S').time()
+            except ValueError:
+                return None
```

Comment:
WDYT about:

```python
formats = {'%Y-%m-%d': 'date', '%H:%M': 'time', %H:%M:%S': 'time'}
for time_format, method in formats.items():
    try:
        time_obj = getattr(datetime.strptime(string, time_format), method)
    except ValueError:
        pass
    else:
        return time_obj
```

Reply:
I tried to avoid passing on an exception, but I can't say I'm happy with how messy it became. I'll implement your version.